[RDY] Command ranges (was PR #1793)

fmoralesc opened the pull request:
This ports patches related to command counts and ranges, which extend vim's command language.
- [x] 7.4.530
- [x] 7.4.531
- [x] 7.4.535
- [x] 7.4.539
- [x] 7.4.542
- [x] 7.4.546
- [x] 7.4.561
- [x] 7.4.565
- [x] 7.4.566
- [x] 7.4.568
- [x] 7.4.572
- [x] 7.4.580
- [x] T.4.584 _NA_
- [x] 7.4.585
- [x] 7.4.588
- [x] 7.4.595

fmoralesc committed:
vim-patch:7.4.530-531

Specify different kinds of counts for commands.
Updated ex commands generator scripts.

Includes fixes to comments from patch 7.4.531

Original message:

Problem:    Many commands take a count or range that is not using line
            numbers.
Solution:   For each command specify what kind of count it uses.  For
            windows, buffers and arguments have "$" and "." have a
relevant
            meaning. (Marcin Szamotulski)

https://code.google.com/p/vim/source/detail?r=v7-4-530
https://code.google.com/p/vim/source/detail?r=v7-4-531

Add legacy tests for 7.4.530

https://code.google.com/p/vim/source/detail?r=1e6d87a36dcdca231721dde8cbbc26610fb3df27

fmoralesc committed:
vim-patch:7.4.539

Patch 7.4.539 (after 7.4.530)
Problem:    Crash when computing buffer count.  Problem with range for
            user commands.  Line range wrong in Visual area.
Solution:   Avoid segfault in compute_buffer_local_count().  Check for
            CMD_USER when checking type of range. (Marcin Szamotulski)

https://code.google.com/p/vim/source/detail?name=v7-4-539

fmoralesc committed:
vim-patch:7.4.535

Patch 7.4.535 (after 7.4.530)
Problem:    Can't build with tiny features.
Solution:   Add #ifdefs and skip a test.

https://code.google.com/p/vim/source/detail?name=v7-4-535

This is also a refactor of some code in 7.4.530.

fmoralesc committed:
vim-patch:7.4.561

Patch 7.4.561
Problem:    Ex range handling is wrong for buffer-local user commands.
Solution:   Check for CMD_USER_BUF. (Marcin Szamotulski)

https://code.google.com/p/vim/source/detail?name=v7-4-561

fmoralesc committed:
vim-patch:7.4.542

Problem:    Using a range for window and buffer commands has a few
            problems.
            Cannot specify the type of range for a user command.
Solution:   Add the -addr argument for user commands.  Fix problems.
            (Marcin Szamotulski

https://code.google.com/p/vim/source/detail?name=v7-4-542

fmoralesc committed:
vim-patch:7.4.565

Problem:    Ranges for arguments, buffers, tabs, etc. are not checked to
            be valid but limited to the maximum.  This can cause the
wrong
            thing to happen.
Solution:   Give an error for an invalid value. (Marcin Szamotulski)
            Use windows range for ":wincmd".

https://code.google.com/p/vim/source/detail?r=v7-4-565

fmoralesc committed:
vim-patch:7.4.566

Problem:    :argdo, :bufdo, :windo and :tabdo don't take a range.
Solution:   Support the range. (Marcin Szamotulski)

https://code.google.com/p/vim/source/detail?r=v7-4-566

fmoralesc committed:
vim-patch:7.4.568

Problem:    Giving an error for ":0wincmd w" is a problem for some
            plugins.
Solution:   Allow the zero in the range. (Marcin Szamotulski)

https://code.google.com/p/vim/source/detail?r=v7-4-568

fmoralesc committed:
vim-patch:7.4.546

Problem:    Repeated use of vim_snprintf() with a number.
Solution:   Move these vim_snprintf() calls into a function.

https://code.google.com/p/vim/source/detail?name=v7-4-546

fmoralesc committed:
vim-patch:7.4.585

Problem:    Range for :bdelete does not work. (Ronald Schild)
Solution:   Also allow unloaded buffers.

https://code.google.com/p/vim/source/detail?name=v7-4-585

fmoralesc committed:
vim-patch:7.4.588

Problem:    ":0argedit foo" puts the new argument in the second place
instead of the first.
Solution:   Adjust the range type. (Ingo Karkat)

https://code.google.com/p/vim/source/detail?name=v7-4-588

fmoralesc committed:
vim-patch:7.4.595

Problem:    The test_command_count test fails when using Japanese.
Solution:   Force the language to C. (Hirohito Higashi)

https://code.google.com/p/vim/source/detail?name=v7-4-595

Pyrohh commented on runtime/doc/windows.txt:
```diff
@@ -263,28 +263,54 @@ left of the Vim window.
+		    :-quit  " quit the previous window
+		    :+quit  " quit the next window
+		    :+2quit " will also work as expected
+<
```
stray `<`?

fmoralesc replied:
It's like that in the original. It signals the commands section end.

Pyrohh replied:
Ah okay, please disregard.

On March 4, 2015 6:43:30 PM EST, Felipe Morales <notifications@github.com> wrote:
>>  
>> +		When 'hidden' is set, and there is only one window for the
>> +		current buffer, it becomes hidden. When 'hidden' is not set, 
>> +		and there is only one window for the current buffer, and the
>> +		buffer was changed, the command fails.
>> +		(Note: CTRL-Q does not 
>> +		work on all terminals). If [count] is greater than
>> +		the last window number the last window will be closed: >
>> +		    :1quit  " quit the first window
>> +		    :$quit  " quit the last window
>> +		    :9quit  " quit the last window
>> +			    " if there are less than  windows opened
>> +		    :-quit  " quit the previous window
>> +		    :+quit  " quit the next window
>> +		    :+2quit " will also work as expected
>> +<
>
>It's like that in the original. It signals the commands section end.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/neovim/neovim/pull/2041/files#r25826711

Pyrohh commented on runtime/doc/windows.txt:
```diff
@@ -263,28 +263,54 @@ left of the Vim window.
+		current buffer, it becomes hidden. When 'hidden' is not set, 
+		and there is only one window for the current buffer, and the
+		buffer was changed, the command fails.
+		(Note: CTRL-Q does not 
```
This could probably be joined with the previous line up to the closing paren, like so:
```
(Note: CTRL-Q does not work on all terminals)
```

glts replied:
:+1: agreed.

Pyrohh commented on runtime/doc/windows.txt:
```diff
@@ -679,8 +715,10 @@ can also get to them with the buffer list commands, like ":bnext".
-:bufdo[!] {cmd}		Execute {cmd} in each buffer in the buffer list.
-			It works like doing this: >
+:[range]bufdo[!] {cmd}	Execute {cmd} in each buffer in the buffer list or if 
+                        [range[ is given only for buffers for which their
```
`[range[` -> `[range]`

fmoralesc replied:
:+1:

Pyrohh commented on src/nvim/ex_cmds.lua:
```diff
@@ -28,2610 +28,3136 @@ local FILES      = bit.bor(XFILE, EXTRA)
+local ADDR_LINES            =  0
+local ADDR_WINDOWS          =  1
+local ADDR_ARGUMENTS        =  2
+local ADDR_LOADED_BUFFERS   =  3
```
Perhaps `ADDR_BUFFERS_LOADED`? for consistency with `ADDR_BUFFERS` below.

Pyrohh commented on src/nvim/ex_cmds_defs.h:
```diff
@@ -71,6 +71,14 @@
+#define ADDR_LINES              0
+#define ADDR_WINDOWS            1
+#define ADDR_ARGUMENTS          2
+#define ADDR_LOADED_BUFFERS     3
```
(copied from previous comment) Perhaps `ADDR_BUFFERS_LOADED`? for consistency with ADDR_BUFFERS` below.

fmoralesc replied:
Not sure about this name change. I would prefer to make it sense as english instead.

Pyrohh replied:
Okay, fair enough. I have no strong preference either way.

On March 4, 2015 10:08:10 PM EST, Felipe Morales <notifications@github.com> wrote:
>> @@ -71,6 +71,14 @@
>>  #define WORD1   (EXTRA | NOSPC) /* one extra word allowed */
>>  #define FILE1   (FILES | NOSPC) /* 1 file allowed, defaults to
>current file */
>>  
>> +// values for cmd_addr_type 
>> +#define ADDR_LINES              0
>> +#define ADDR_WINDOWS            1
>> +#define ADDR_ARGUMENTS          2
>> +#define ADDR_LOADED_BUFFERS     3
>
>Not sure about this name change. I would prefer to make it sense as
>english instead. 
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/neovim/neovim/pull/2041/files#r25836582

Pyrohh commented on src/nvim/ex_docmd.c:
```diff
@@ -3107,12 +3312,11 @@ skip_range (
-    int to_other_file              /* flag: may jump to other file */
-)
+static linenr_T get_address(char_u **ptr,
+                            int addr_type,  // flag: one of ADDR_LINES, ...
```
Why the ellipsis?

Pyrohh replied:
Wait, I can't read.

elmart commented:
This is very much valuable work, but it's not getting attention by reviewers, probably because of being too big (hard to review). Could you try splitting it up into several smaller PRs, or are commits interrelated in such a way that they can't be separated?

fmoralesc commented:
First of all, I'm sorry I've been unable to work on this lately. Life seems
to have catched up hard with me.

As I mentioned in the previous thread, it *could* be best to squash all the
commits in a single unit: some of the changes are basically refactorings of
the initial versions (some repeated code is expanded into a function,
RANGE_BUFFERS becomes RANGE_LOADED_BUFFERS, there's a conditional that kept
been rewritten because of corner cases, iirc), and others are merely
expansions. It's not *terribly* intertwined, though, and I could split it
up in chunks if preferred.




On Fri, Mar 20, 2015 at 7:17 AM, Eliseo Martínez <notifications@github.com>
wrote:

> This is very much valuable work, but it's not getting attention by
> reviewers, probably because of being too big (hard to review). Could you
> try splitting it up into several smaller PRs, or are commits interrelated
> in such a way that they can't be separated?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/neovim/neovim/pull/2041#issuecomment-83976599>.
>

elmart commented:
As for squashing, I think it's best to maintain separated commits for separated upstream patches. So, I'd split this into 3/4 PRs of 4/5 commits each. I bet you'll get more reviews that way. If later commits depend on previous ones having being applied, you'd have to produce those PR's sequentially, though.

If you prefer to leave as it is, well, it's ok. I would try to review this soon. But I'm saying all this because I've already tried a couple of times, and quitted, because feeling not to have the time to do it all at in one session.

fmoralesc commented:
I understand, it is also a pain to fix stuff over such a long series of
commits. the issue with not squashing is some of the patches contain fixes
for stuff the reviewers can find. last review, for example, found some of
those. I'll try to untangle the squashabe threads, and keep you updated.

glts commented:
Fantastic work @fmoralesc! If you need help in any way, please say so, I would love to see this in master.

fmoralesc commented:
A general review would be helpful ;) Please take notice some issues with earlier patches get solved in later ones, so it's best to take a quick look to the complete series before pointing out anything in particular. I'll try to rebase this on master when I get home (tomorrow, most probably), so I can see what needs more work. I'm feeling so bad about leaving this aside for so long!

glts commented:
Thanks for the heads up. Then I suppose I'll wait with reviewing until after the rebase (but *don't* feel pressured! there's no hurry). And there's no need to feel bad obviously :)

oni-link commented on runtime/doc/tabpage.txt:
```diff
@@ -110,12 +119,18 @@ something else.
+		this is a "safe" command. >
+                    :tabclose   " close the current tab page
 
+:{count}tab[lose][!]
```
Missing character `c` after `tab`.

fmoralesc replied:
:+1:

oni-link commented on runtime/doc/tabpage.txt:
```diff
@@ -110,12 +119,18 @@ something else.
+		above. >
+		    :-tabclose      " close the previous tab page
+		    :+tabclose      " close the next tab page
+		    :1tabclose      " close teh first tab page
```
`teh`->`the`

fmoralesc replied:
:+1:

oni-link commented on runtime/doc/tabpage.txt:
```diff
@@ -124,7 +139,16 @@ something else.
+
+:{count}tabo[nly][!]
+		Close all tab pages except the {count}th one. >
+		    :.tabonly " one
```
`close all tab pages except the current one`?

fmoralesc replied:
:+1:

oni-link commented on runtime/doc/tabpage.txt:
```diff
@@ -176,8 +200,15 @@ REORDERING TAB PAGES:
-		page is made the last one.
-
+		page is made the last one. >
+		    :-tabmove   " move the tab page to the left
```
`:-tabmove` does nothing in nvim.

glts replied:
Vim does have the same behaviour/bug at tag 7.4.595, so we could declare fixing it to be out of scope (of this PR).

fmoralesc replied:
Is that fixed in vim's latest version?

glts replied:
No, bug is also in upstream Vim, tested with 7.4.694.

fmoralesc replied:
OK,then, I'll review the suggestions and clean up. Thanks a lot for taking
a look 😉

chrisbra replied:
That's a current bug that is mentioned at vim-dev currently.

oni-link commented on runtime/doc/tabpage.txt:
```diff
@@ -176,8 +200,15 @@ REORDERING TAB PAGES:
-
+		page is made the last one. >
+		    :-tabmove   " move the tab page to the left
+		    :tabmove    " move the tab page to the right
```
`:tabmove` does the same as `:$tabmove`.

glts replied:
Same here, bug is also in Vim 7.4.595.

glts replied:
Same here again, bug is also in current Vim 7.4.694.

oni-link commented on runtime/doc/windows.txt:
```diff
@@ -263,28 +263,54 @@ left of the Vim window.
+		    :1quit  " quit the first window
+		    :$quit  " quit the last window
+		    :9quit  " quit the last window
+			    " if there are less than  windows opened
```
`... 9 windows opened`

oni-link replied:
Does not seem to work as described. With less than 9 windows an invalid range error is given.

justinmk replied:
I believe some of these issues were addressed or discussed in other patches on vim_dev.

fmoralesc replied:
Maybe in some of the patches since this set? I'll take a look.

fmoralesc commented:
Rebased the commit set over master and included some of the latest comments. When reviewing, please consider some of the patches are actually fixing issues over previous ones. Likewise, I think it should be better to refer to the tip of the branch for style comments.

glts commented on runtime/doc/editing.txt:
```diff
@@ -801,8 +801,9 @@ current window.  The two windows then share this list, until one of them uses
-			It works like doing this: >
+:[range]argdo[!] {cmd}	Execute {cmd} for each file in the argument list or
+                        if [range] is specified only for arguments in that
+			range. It works like doing this: >
```
Mixed tabs and spaces. Double space after sentence punctuation.

I'll add a note about whitespace/formatting at the bottom.

glts commented on runtime/doc/tabpage.txt:
```diff
@@ -61,16 +61,25 @@ In the GUI tab pages line you can use the right mouse button to open menu.
 		Open a new tab page with an empty window, after the current
-		tab page.  For [count] see |:tab| below.
+		tab page.  If [count] is given the new tab page appears after 
+		the tabpage [count] otherwise the new tab page will appear
```
Here and below `tabpage` -> `tab page` for consistency.

Pyrohh replied:
@fmoralesc (small reminder about this, just so you don't forget to address it, or at least respond to it)

glts commented on runtime/doc/tabpage.txt:
```diff
@@ -124,7 +139,16 @@ something else.
+		    :-tabonly " close all tab pages except the previous one
+		    :+tabonly " close all tab pages except the next one
+		    :1tabonly " close all tab pages except the first one
+		    :$tabonly " close all tab pages except the last one.
```
Perhaps a little more space after `tabonly` like in the patch?

glts commented on runtime/doc/windows.txt:
```diff
@@ -263,28 +263,54 @@ left of the Vim window.
 Closing a window
```
Some trailing whitespace in this file.

glts commented on runtime/doc/windows.txt:
```diff
@@ -263,28 +263,54 @@ left of the Vim window.
-		current buffer, and the buffer was changed, the command fails.
-		(Note: CTRL-Q does not work on all terminals)
+		Without {count}: Quit the current window.  If {count} is 
+		given quit the {count} window
```
`window.`

glts commented on runtime/doc/windows.txt:
```diff
@@ -263,28 +263,54 @@ left of the Vim window.
+:q[uit]!
+:{count}q[uit]!
+		Without {count}: Quit the current window.  If {count} is 
+		given quit the {count} window
```
`window.`

Pyrohh replied:
Also, this sentence is structured oddly IMO, given the colon. Perhaps:
```
Without {count}, quit the current window.  If {count} is 
given, quit the {count} window.
```

Still, that doesn't address how it sounds when you substitute a number for `{count}`, e.g.

```
If 5 is given quit the 5 window.
```
should it be `close the next {count} windows`?

glts commented on runtime/doc/windows.txt:
```diff
@@ -679,8 +715,10 @@ can also get to them with the buffer list commands, like ":bnext".
-			It works like doing this: >
+:[range]bufdo[!] {cmd}	Execute {cmd} in each buffer in the buffer list or if 
+                        [range[ is given only for buffers for which their
+                        buffer name is in the [range]. It works like doing 
```
`buffer number`, not name.

glts commented on runtime/doc/windows.txt:
```diff
@@ -992,8 +1030,11 @@ list of buffers. |unlisted-buffer|
-		cleared.
-
+		cleared. Examples: >
+                    :.,$-bdelete   "delete buffers from the current one to
```
Space after `"`.

glts commented on runtime/doc/windows.txt:
```diff
@@ -1016,8 +1057,11 @@ list of buffers. |unlisted-buffer|
-		unless you know what you are doing.
-
+		unless you know what you are doing. Examples: >
+                    :.+,$wipeout    " wipe out all buffers after the current
```
Should read `bwipeout`.

glts commented:
I noticed a few whitespace/formatting problems while doing the doc review. `:set list` is very useful here.

- trailing whitespace: `runtime/doc/windows.txt` especially has some trailing whitespace
- indentation: Vim doc consistently uses real tabs to indent. This is made possible via the modeline `ts=8` setting present in every help file. I think it's best to follow this convention for now, and not mix tabs/spaces as done here.
- line length should not be greater than 78. This is Vim doc convention and again codified in the modeline setting `tw=78` present in all help files.
- double spacing after sentence punctuation. Help files use double spacing after sentence punctuation. For the time being I think it's best to follow the convention and maintain consistency with the existing help corpus.

glts commented:
and if you think I'm reading to closely, just ignore :) <sub>I can be a bit petty</sub>

fmoralesc commented:
it's weird that my vim doesn't seem to respect the modeline settings, then.
I'll take a look.

fmoralesc commented:
No problem. I'm worse when proofreading stuff written by others, but I hate
editing myself :p

splinterofchaos commented:
What a massive PR! I have to admire your tenacity. (Took a long time to review.)

glts commented:
One general style note: we always use curly brackets, see style guide http://neovim.org/develop/style-guide.xml#Conditionals. In new code introduced by a patch please remember to add them whenever possible.

glts commented:
What a humongous changeset! I'm exhausted. I finally managed to look at everything. I think it looks good.

If you could use any kind of support do say so. Let's get this over and done with :)

fmoralesc commented:
Rebased again and included the fixes from the last batch of reviews, save one thing (the mixup between patch 565 and 572).

Pyrohh commented on runtime/doc/editing.txt:
```diff
@@ -801,8 +801,9 @@ current window.  The two windows then share this list, until one of them uses
-			It works like doing this: >
+:[range]argdo[!] {cmd}	Execute {cmd} for each file in the argument list or
+			if [range] is specified only for arguments in that
+			range. It works like doing this: >
```
Should be two spaces between period.

Pyrohh commented on runtime/doc/map.txt:
```diff
@@ -1358,6 +1358,20 @@ Possible attributes are:
 specified.
 
+                                                  *E889* *:command-addr*
+It is possible that the special characters in the range like ., $ or % which
```
Perhaps the mentioned special characters should be surrounded by double quotes or backtics. Stuff like `range like ., $` are odd to read IMO.

fmoralesc replied:
Good idea.

Pyrohh commented on runtime/doc/tabpage.txt:
```diff
@@ -61,16 +61,25 @@ In the GUI tab pages line you can use the right mouse button to open menu.
+		tab page.  If [count] is given the new tab page appears after 
+		the tabpage [count] otherwise the new tab page will appear
+		after the current one. >
+		    :tabnew     " opens tabpage after the current one
```
I've never liked how in the vim docs, indented blocks such as this aren't separated between the above paragraph, like this:
```
after the current one. >

		    :tabnew     " opens tabpage after the current one
```

What do you think?

fmoralesc replied:
I'm neutral on this one. Perhaps it should be included in a further PR to clean up the whole thing.

Not sure how that would impact merging changes to the documentation from vim, though.

Pyrohh replied:
There are only a few instances of it in this PR, so I'd just do those now. I don't think it's worth it to make a PR just for this, well maybe... idk. It's up to you.

fmoralesc replied:
I was thinking on a PR fixing this in every documentation file... 

But no, I think after this PR I might go small for my next contrib :p

Pyrohh commented on runtime/doc/tabpage.txt:
```diff
@@ -191,8 +222,10 @@ clarification what +N means in this context see |[range]|.
-:tabd[o] {cmd}	Execute {cmd} in each tab page.
-		It works like doing this: >
+:[range]tabd[o] {cmd}
+        	Execute {cmd} in each tab page or if [range] is given only in
```
comma after `tab page`?

fmoralesc replied:
:+1:

fmoralesc replied:
Maybe

~~~
		Execute {cmd} in each tab page or, if [range] is given, only
		in tabpages which tab page number is in the [range]. It works
~~~

Pyrohh replied:
:+1:

fmoralesc commented:
Applied @Pyrohh's fixes and fixed the mixup with 565 and 572. Should be good now.

fmoralesc commented:
The build is consistently failing on the OSX Travis builds, if someone could give a look it would be appreciated.

splinterofchaos commented:
Some of these errors,
```
/Users/travis/build/neovim/neovim/test/unit/garray_spec.lua:134: missing declaration for symbol 'malloc'
```
I was getting in another PR, so there may be a problem with the units themselves.

Pyrohh commented on src/nvim/window.c:
```diff
@@ -488,6 +488,107 @@ do_window (
+static void cmd_with_count(char *cmd, char_u *bufp, size_t bufsize,
+                           uint32_t Prenum)
+{
+  size_t len = STRLEN(cmd);
```
Wow! I haven't seen `STRLEN` without being casted to `int` in years...

fwalch replied:
Since `cmd` is a `char *`, this could actually be plain `strlen`.

fmoralesc replied:
I wake up at 6am on a sunday because of a fire truck siren, and sure enough, it was a code review from @Pyrohh ...

:+1:

Pyrohh replied:
What kind of goofball lives inside a fire truck?

On April 19, 2015 5:23:51 AM EDT, Felipe Morales <notifications@github.com> wrote:
>> +    case 'b':
>> +    case Ctrl_B:
>> +    case 'p':
>> +    case Ctrl_P:
>> +    case '=':
>> +    case CAR:
>> +      /* no count */
>> +      eap->addr_type = 0;
>> +      break;
>> +  }
>> +}
>> +
>> +static void cmd_with_count(char *cmd, char_u *bufp, size_t bufsize,
>> +                           uint32_t Prenum)
>> +{
>> +  size_t len = STRLEN(cmd);
>
>I wake up at 6am on a sunday because of a fire truck siren, and sure
>enough, it was a code review from @Pyrohh ...
>
>:+1:
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/neovim/neovim/pull/2041/files#r28651168

fmoralesc replied:
![](http://s1.b3ta.com/host/creative/65077/1316423102/FiremanSad.gif)

Pyrohh commented on src/nvim/window.c:
```diff
@@ -488,6 +488,107 @@ do_window (
+}
+
+static void cmd_with_count(char *cmd, char_u *bufp, size_t bufsize,
+                           uint32_t Prenum)
```
I wonder why this parameter is capitalized...

Pyrohh replied:
nevermind

oni-link commented on src/nvim/ex_docmd.c:
```diff
@@ -1335,25 +1419,108 @@ static char_u * do_one_cmd(char_u **cmdlinep,
+      // ea.addr_type for user commands is set by find_ucmd
+    }
+  // :wincmd range depends on the argument
+  } else if (ea.cmdidx == CMD_wincmd) {
```
This looks like unreachable code.

fmoralesc replied:
Was it because of the ea.cmdidx >= 0 check (which I just removed), or something else? I don't see it.

oni-link replied:
My fault. This line can be reached, but the condition should always be false, because the previous `if` also matches this condition. `!IS_USER_CMDIDX(ea.cmdidx)` is `(int)(ea.cmdidx)>=0` and `CMD_wincmd` should be not negative.

fmoralesc replied:
Ah , I see now. I really botched this one up, this section should be 
~~~ c
  // ea.addr_type for user commands is set by find_ucmd
  if (!IS_USER_CMDIDX(ea.cmdidx)) {
    if (ea.cmdidx != CMD_SIZE) {
      ea.addr_type = cmdnames[(int)ea.cmdidx].cmd_addr_type;
    } else {
      ea.addr_type = ADDR_LINES;
    }
    // :wincmd range depends on the argument
    if (ea.cmdidx == CMD_wincmd) {
      get_wincmd_addr_type(skipwhite(p), &ea);
    }
  }
~~~

oni-link commented on src/nvim/ex_docmd.c:
```diff
@@ -1335,25 +1419,108 @@ static char_u * do_one_cmd(char_u **cmdlinep,
    */
 
+  if (!IS_USER_CMDIDX(ea.cmdidx)) {
+    if (ea.cmdidx != CMD_SIZE && ea.cmdidx >= 0) {
```
`ea.cmdidx >= 0` is always true at this point.

fmoralesc replied:
I added that because the build complained about this. Maybe something else changed in the meanwhile, because I don't see that issue anymore when I remove it.

glts commented on src/nvim/ex_docmd.c:
```diff
@@ -1335,25 +1419,109 @@ static char_u * do_one_cmd(char_u **cmdlinep,
+      ea.addr_type = ADDR_LINES;
+    }
+    // :wincmd range depends on the argument
+    if (ea.cmdidx == CMD_wincmd) {
```
Patch 7.4.580 added a `&& p != NULL` check here. Left out on purpose?

fmoralesc replied:
I bundled the skipwhite part of that patch (don't remember why, but I probably didn't notice 580 at all) into 572. I'll add this check there and change the commit description.

fmoralesc committed:
vim-patch:7.4.580

Problem:    ":52wincmd v" still gives an invalid range error. (Charles
            Campbell)
Solution:   Skip over white space.

https://code.google.com/p/vim/source/detail?name=v7-4-580

fmoralesc committed:
vim-patch:7.4.572

Problem:    Address type of :wincmd depends on the argument.
Solution:   Check the argument.

https://code.google.com/p/vim/source/detail?r=v7-4-572

mhinz commented:
LGTM

justinmk commented on src/nvim/ex_docmd.c:
```diff
@@ -1065,16 +1066,178 @@ void * getline_cookie(LineGetter fgetline,
+#define LAST_TAB_NR current_tab_nr(NULL)
+
+/*
+* Figure out the address type for ":wnncmd".
```
should be ":wincmd". I'll fix it it before merging...

justinmk replied:
Should say ":wincmd". Also might as well fix https://github.com/neovim/neovim/pull/2041/files#r28235543

Then after verifying the tests are working, let's merge this...

fmoralesc replied:
:+1:

oni-link commented on src/nvim/window.c:
```diff
@@ -487,6 +487,17 @@ do_window (
+{
+  size_t len = strlen(cmd);
+
+  STRCPY(bufp, cmd);
```
Is `len` always shorter than `bufsize`? Using `xstrlcpy()` would be safer.

fmoralesc replied:
Ah, I thought the size was the size of the `dst` value.

fmoralesc commented:
OK. Included the latest comments and cleaned up. Travis is marked as failed because one of the dependencies failed to download in the linux-gcc build, but that's not a big deal.

oni-link commented on src/nvim/window.c:
```diff
@@ -487,6 +487,17 @@ do_window (
+{
+  size_t len = strlen(cmd);
+
+  xstrlcpy((char *)bufp, cmd, len + 1);
```
`bufp` has size `bufsize`, so it should be something like this
```c
size_t len = xstrlcpy((char *)bufp, cmd, bufsize);
if (Prenum > 0 && len < bufsize) {
...
}
```

Pyrohh commented on src/nvim/ex_docmd.c:
```diff
@@ -4182,6 +4602,21 @@ static int uc_add_command(char_u *name, size_t name_len, char_u *rep,
 
+
+static struct {
+    int expand;
```
2 space indentation

Pyrohh commented on src/nvim/ex_docmd.c:
```diff
@@ -4182,6 +4602,21 @@ static int uc_add_command(char_u *name, size_t name_len, char_u *rep,
+{
+    {ADDR_ARGUMENTS, "arguments"},
+    {ADDR_LINES, "lines" },
+    {ADDR_LOADED_BUFFERS, "loaded_buffers"},
```
2 space indentation

Pyrohh commented:
@fmoralesc might be a good idea to rebase before you repush.

fmoralesc commented:
Done.

justinmk commented on src/nvim/window.c:
```diff
@@ -487,6 +487,16 @@ do_window (
+  size_t len = xstrlcpy((char *)bufp, cmd, bufsize);
+
+  if (Prenum > 0 && len < bufsize) {
+    vim_snprintf((char *)bufp + len, bufsize - len, "%ld", Prenum);
```
Use a portable format specifier (like the old Neovim version, here: https://github.com/neovim/neovim/pull/2041/files#diff-40d9eb21ee75b3df148c827ec4d4310aL129 ). `Prenum` is `long` so `PRId64` is the format specifier for that. E.g.:

```c
  if (Prenum > 0 && len < bufsize) {
    vim_snprintf((char *)bufp + len, bufsize - len, "%" PRId64, Prenum);
```

Why does `cmd_with_count()` declare `Prenum` as `uint32_t` instead of `long`?


See also: https://github.com/neovim/neovim/pull/1729#discussion_r22436056

fmoralesc replied:
No reason. Sorry for the oversight!

justinmk replied:
No problem. Many thanks for this patch series!

fmoralesc replied:
Nah, thank you all for the patience ;)

justinmk commented:
The commit message `vim-patch: 7.4.595` won't be picked up by the vim-patch report because there's a space. 

LGTM after that is fixed. @oni-link are you finished reviewing?

Pyrohh commented on src/nvim/ex_cmds_defs.h:
```diff
@@ -71,6 +71,14 @@
 #define WORD1   (EXTRA | NOSPC) /* one extra word allowed */
 #define FILE1   (FILES | NOSPC) /* 1 file allowed, defaults to current file */
 
+// values for cmd_addr_type 
```
trailing whitespace here sonny :D

fmoralesc replied:
Of course there is trailing white space :0 I'll fix this once I get home
later today.

Pyrohh commented on runtime/doc/editing.txt:
```diff
@@ -801,8 +801,9 @@ current window.  The two windows then share this list, until one of them uses
 						*:argdo*
-:argdo[!] {cmd}		Execute {cmd} for each file in the argument list.
-			It works like doing this: >
+:[range]argdo[!] {cmd}	Execute {cmd} for each file in the argument list or
```
maybe comma after `list`

fmoralesc replied:
Turned it into 
~~~
:[range]argdo[!] {cmd}	Execute {cmd} for each file in the argument list or,
			if [range] is specified, only for arguments in that
~~~

fmoralesc replied:
Nevermind the spaces ;)

Pyrohh commented on runtime/doc/tabpage.txt:
```diff
@@ -110,12 +119,18 @@ something else.
+		this is a "safe" command. >
+                    :tabclose   " close the current tab page
 
+:{count}tabc[lose][!]
```
What's the syntactic purpose of using braces here as opposed to square brackets? It seems like they're being arbitrarily interchanged (for example, `:[range]argdo[!]` above, where it seems like the brackets mean an optional part of the command.

justinmk replied:
Yeah, I think that slipped by Bram. `[count]` is standard.

Pyrohh commented on runtime/doc/tabpage.txt:
```diff
@@ -124,7 +139,16 @@ something else.
+		never abandoned, so changes cannot get lost. >
+		    :tabonly " close all tab pages except the current one
+
+:{count}tabo[nly][!]
```
same regarding brackets

Pyrohh commented on runtime/doc/windows.txt:
```diff
@@ -263,28 +263,54 @@ left of the Vim window.
 ----------------
 
+:q[uit]
+:{count}q[uit]
```
same regarding brackets (there's more, but I don't want to flood your inbox). If this is inconsistent throughout, perhaps a separate PR should deal wtih it.

Pyrohh commented on runtime/doc/tabpage.txt:
```diff
@@ -191,8 +222,10 @@ clarification what +N means in this context see |[range]|.
-:tabd[o] {cmd}	Execute {cmd} in each tab page.
-		It works like doing this: >
+:[range]tabd[o] {cmd}
+		Execute {cmd} in each tab page or, if [range] is given, only
```
maybe the comma should go before `or`?

Pyrohh commented on runtime/doc/tabpage.txt:
```diff
@@ -191,8 +222,10 @@ clarification what +N means in this context see |[range]|.
-		It works like doing this: >
+:[range]tabd[o] {cmd}
+		Execute {cmd} in each tab page or, if [range] is given, only
+		in tabpages which tab page number is in the [range]. It works
```
2 spaces after period.

Pyrohh commented on runtime/doc/windows.txt:
```diff
@@ -263,28 +263,54 @@ left of the Vim window.
+		exit Vim.
 
+		When 'hidden' is set, and there is only one window for the
+		current buffer, it becomes hidden. When 'hidden' is not set,
```
2 spaces after period

Pyrohh replied:
I see why people wanted to rewrite the docs in markdown, it would be much more consistent, plus we could use semantic newlines for easier diffs.

Pyrohh commented on runtime/doc/windows.txt:
```diff
@@ -263,28 +263,54 @@ left of the Vim window.
+		and there is only one window for the current buffer, and the
+		buffer was changed, the command fails.
+		(Note: CTRL-Q does not
+		work on all terminals). If [count] is greater than
```
2 spaces after period, and the `Note:` should be on one line probably.

Pyrohh commented on runtime/doc/windows.txt:
```diff
@@ -263,28 +263,54 @@ left of the Vim window.
+:q[uit]!
+:{count}q[uit]!
+		Without {count}: Quit the current window.  If {count} is
+		given quit the {count} window
```
missing period

Pyrohh commented on runtime/doc/windows.txt:
```diff
@@ -263,28 +263,54 @@ left of the Vim window.
+		Without {count}: Close the current window. If given close the
+		{count} window.
+
+		When the 'hidden' option is set, or when the buffer was changed
```
Maybe just `when 'hidden' is set`; this seems more idiomatic IMO, as it should be obvious that it's an option given the single quotes.

Pyrohh commented on runtime/doc/windows.txt:
```diff
@@ -297,29 +323,38 @@ CTRL-W CTRL-C						*CTRL-W_CTRL-C*
+:hid[e]
+:{count}hid[e]
+		Quit the current window, unless it is the last window on the
+		screen.  For {count} see |:quit| command.
```
Maybe just `... see |:quit|.`

Pyrohh commented on runtime/doc/windows.txt:
```diff
@@ -297,29 +323,38 @@ CTRL-W CTRL-C						*CTRL-W_CTRL-C*
+		screen.  For {count} see |:quit| command.
+
+		The buffer becomes hidden (unless there is another window
+		editing it or 'bufhidden' is "unload" or "delete"). If the
```
unload and delete could be surrounded by bacticks instead.

Pyrohh commented on runtime/doc/windows.txt:
```diff
@@ -297,29 +323,38 @@ CTRL-W CTRL-C						*CTRL-W_CTRL-C*
-:on[ly][!]	Make the current window the only one on the screen.  All other
-		windows are closed.
+		Make the current window the only one on the screen.  All other
+		windows are closed. For {count} see |:quit| command.
```
`see |:quit|.`

Pyrohh commented on runtime/doc/windows.txt:
```diff
@@ -679,8 +715,10 @@ can also get to them with the buffer list commands, like ":bnext".
+:[range]bufdo[!] {cmd}	Execute {cmd} in each buffer in the buffer list or if
+			[range[ is given only for buffers for which their
+			buffer name is in the [range]. It works like doing
+                        this: >
```
should be tabs, not spaces

Pyrohh commented on runtime/doc/windows.txt:
```diff
@@ -679,8 +715,10 @@ can also get to them with the buffer list commands, like ":bnext".
-			It works like doing this: >
+:[range]bufdo[!] {cmd}	Execute {cmd} in each buffer in the buffer list or if
+			[range[ is given only for buffers for which their
+			buffer name is in the [range]. It works like doing
```
2 spaces after period

elmart commented:
Just one comment: Time ago we talked about two spaces after period and decided against that. There are still lots of them in untouched comments, but we've been removing it in modified ones.

Pyrohh commented on src/nvim/ex_docmd.c:
```diff
@@ -1131,7 +1295,7 @@ static char_u * do_one_cmd(char_u **cmdlinep,
   for (;; ) {
     /*
-     * 1. skip comment lines and leading white space and colons
+     * 1. Skip comment lines and leading white space and colons
```
missing period

justinmk commented:
Right, please don't *enforce* 2 spaces after period.

Pyrohh commented on src/nvim/ex_docmd.c:
```diff
@@ -3294,15 +3664,65 @@ static void ex_script_ni(exarg_T *eap)
+      case ADDR_LINES:
+        if (!(eap->argt & NOTADR)
+            && eap->line2 > curbuf->b_ml.ml_line_count
+            + (eap->cmdidx == CMD_diffget)) {
```
Since this line is part of the `&&`, its start should line up with `e` from `eap->line2 ...` above

Pyrohh commented on src/nvim/ex_docmd.c:
```diff
@@ -4182,6 +4602,21 @@ static int uc_add_command(char_u *name, size_t name_len, char_u *rep,
+} addr_type_complete[] =
+{
+  {ADDR_ARGUMENTS, "arguments"},
+  {ADDR_LINES, "lines" },
```
extra space after second quotation mark.

Pyrohh commented:
@elmart @justinmk Huh, no-one ever told me. I deleted all comments regarding them.

glts commented:
@elmart @justinmk Please keep in mind that the help docs are among the parts of (Neo)Vim that have the most immediate user visibility. I think a mix of the two styles could be seen as a lack of attention to detail or sloppiness. I am a careful reader and I can say that it would irritate me.

I don't want to distract from this PR so I won't insist here.

elmart commented:
Well, somebody could do a PR replacing all `<period><space><space>` sequences with `<period><sequence>` once and for all, so that people don't get confused on this, and readers are not bothered by that. I'm now busy with other things, but will do if nobody else does before.

justinmk commented:
> I think a mix of the two styles could be seen as a lack of attention to detail or sloppiness. I am a careful reader and I can say that it would irritate me.

I agree and appreciate that. We will do our best to weigh the costs and benefits :)

Pyrohh commented:
@elmart I think it would be better if we just switched to a format which deals with menial things like line wrapping and spaces after periods in a consistent and sane manner. Of course that's a lot of work, but it's better than making such a PR IMO, which has no promise of maintaining that consistency, e.g. when new patches are merged.

Pyrohh commented:
Anyways, that's just an idea. I don't want to distract from this PR anymore.

justinmk commented:
@Pyrohh Markdown has its own problems. There's discussion here: https://github.com/neovim/neovim/issues/329

But it's very low priority, IMO, for the next year or so.

elmart commented:
@Pyrohh I do agree markdown can be better for user doc (not without its problems, as @justinmk said). But that can take a while. What I was proposing was a quick fix for the mentioned problem. In any case, note that markdown could be used for user doc, but you should still have to deal with source code comments, which also have two spaces after period in many places.

fmoralesc commented:
OK, included the latest comments (including removing the double spaces after periods). I'll rebase at the end once you tell me things are OK.

Pyrohh commented:
The tests haven't passed yet, but besides that LGTM.

fmoralesc commented:
Good thing is it always fails differently, it gives the whole ordeal some extra flavor.

Pyrohh commented:
Hah, it's like russian roulette.

fmoralesc commented:
OK, then, I'll rebase to clean up the history and to give Travis another chance to fail on me (at this point, it's personal).

fmoralesc commented:
![it's over](https://i.imgflip.com/kr357.jpg)

Pyrohh commented:
https://i.imgur.com/iYZM3QR.gifv

justinmk commented on src/nvim/ex_docmd.c:
```diff
@@ -1065,16 +1066,178 @@ void * getline_cookie(LineGetter fgetline,
+#define CURRENT_WIN_NR current_win_nr(curwin)
+#define LAST_WIN_NR current_win_nr(NULL)
+#define CURRENT_TAB_NR current_tab_nr(curtab)
+#define LAST_TAB_NR current_tab_nr(NULL)
```
These don't need to be macros in Neovim. We can leave that for a different PR though.

justinmk commented:
If I comment out any (or all) of `test_argument_count.in`, then `make oldtest` still passes.

fmoralesc commented:
@justinmk You probably need to remove the .out files manually... I just tried changing one of the `.ok` files and it catches the error. It's weird though, I assumed the .out files were cleaned up automatically.
~~~
rm -rf test_command_count.failed test.ok test.out X* viminfo
cp test_command_count.ok test.ok
# Sleep a moment to avoid that the xterm title is messed up.
# 200 msec is sufficient, but only modern sleep supports a fraction of
# a second, fall back to a second if it fails.
VIMRUNTIME=../../../runtime; export VIMRUNTIME;  ../../../build/bin/nvim -u unix.vim -U NONE -i viminfo --noplugin -s dotest.in test_command_count.in
# For flaky tests retry one time.
# Check if the test.out file matches test.ok.
12c12
< RangeBuffers 1 5
---
> RangeBuffers 1 4
rm -rf X* test.ok viminfo
rm -rf test_argument_0count.failed test.ok test.out X* viminfo
cp test_argument_0count.ok test.ok
# Sleep a moment to avoid that the xterm title is messed up.
# 200 msec is sufficient, but only modern sleep supports a fraction of
# a second, fall back to a second if it fails.
VIMRUNTIME=../../../runtime; export VIMRUNTIME;  ../../../build/bin/nvim -u unix.vim -U NONE -i viminfo --noplugin -s dotest.in test_argument_0count.in
# For flaky tests retry one time.
# Check if the test.out file matches test.ok.
rm -rf X* test.ok viminfo

Test results:
test_command_count FAILED
TEST FAILURE
Makefile:77: fallo en las instrucciones para el objetivo 'report'
make: *** [report] Error 1
~~~

fmoralesc commented:
OK, so there's two things pending:
* @oni-link found a bug in the handling of `:close`, which is also present in vim, so we might want to treat that separatedly (notify vim_dev, etc.). I don't think I'll fix it at the commit it is introduced, but in a separate commit or PR even.
* There's some instances of `{count}cmd`, which you tell me would be best changed into `[count]cmd`. I found the idiom is used elsewhere in the documentation (see `:help :tabprevious`), so I don't feel like it's worth fussing over it. It is used to convey that what's between the `{}` is variable, instead of just optional. I also think it looks uglier in square brackets, because other items in the same lines also use them, and `{}` help differentiate them. E.g.
~~~  diff
-:{count}tabc[lose][!]
-:tabc[lose][!] {count}
-		Close tab page {count}.  Fails in the same way as `:tabclose`
+:[count]tabc[lose][!]
+:tabc[lose][!] [count]
+		Close tab page [count].  Fails in the same way as `:tabclose`
 		above. >
~~~
So, unless you *really* think these should be changed and give me a good reason for it, I'm inclined not to.

Pyrohh commented:
> There's some instances of {count}cmd, which you tell me would be best changed into [count]cmd. I found the idiom is used elsewhere in the documentation (see :help :tabprevious), so I don't feel like it's worth fussing over it. It is used to convey that what's between the {} is variable, instead of just optional. I also think it looks uglier in square brackets, because other items in the same lines also use them, and {} help differentiate them. E.g.

Fine with me, although it should be made consistent one way or the other (not necessarily in this PR).

justinmk commented:
> {count}cmd, which you tell me would be best changed into [count]cm

No, I was just mentioning it.

When I ran the tests in a local ubuntu VM after a *rebase* and `rm -rf .deps build` (with no `.out` files), then `test_close_count.in` failed consistently, with no local changes (except for the rebase).

I am going to try it after a merge instead of a rebase. Ideally it would be best to convert the tests to lua tests.

fmoralesc commented:
> I am going to try it after a merge instead of a rebase. Ideally it would be best to convert the tests to lua tests.

Agreed. Do you need me to do it in this patchset? I don't feel entirely confident with the testing framework, so if someone could drop in, it would be appreciated.

fmoralesc commented:
@Pyrohh No, I think they should *not* be made consistent. As I said, `{var}` is intended to signal a variable part of the command, whereas `[...]` marks optional stuff. See

    {count}clo[se]

`:countclose` is meaningless, but `:2clo` isn't.

Pyrohh commented:
> No, I think they should not be made consistent

I think I was unclear, as I didn't mean that.

> As I said, {var} is intended to signal a variable part of the command, whereas [...] marks optional stuff.

It doesn't seem like that in at least a few places; see `:helpgrep` and `:vimgrep`, where the brackets around `{pattern}` seem to indicate it's non-optional.

justinmk commented:
> Do you need me to do it in this patchset?

@fmoralesc No, let me see if I can knock it out in the next hour.

fmoralesc commented:
@justinmk Sure. Btw, was it you who restarted the travis build? That red made me sad... ;)

justinmk commented:
@fmoralesc Yes, but dontt worry about the red.

fmoralesc commented:
Rebased again agains master and fixed the commit message issue @oni-link found (in a previous version of the patch, I had merged commits 572 and 580 it seems). 

On the documentation point re curly braces and square brackets, I've been chatting with @Pyrohh 
and we've convened that the best convention is probably to use square brackets for any optional group, and curly braces for placeholder. So, for example,

     [range]windo {cmd}

should become

    [{range}]windo {cmd}

and 

    {count}clo[se][!]

should become

    [{count}]clo[se][!]

Sadly, this will necessitate changes in the help syntax file, so I'll make those changes in a separate PR not to clutter this one.

I'll wait for comments on the bug @oni-link found. See [here](https://github.com/fmoralesc/neovim/commit/e22043becebd0310c12af2af82928fdf1598db11#inlinecommitcomment-10953797)

justinmk commented:
Migrating the tests is taking too long. Instead I was able to verify the individual legacy tests like so:

```sh
cd src/nvim/testdir
make clean && TESTNUM=_command_count make
make clean && TESTNUM=_argument_0count make
make clean && TESTNUM=_argument_count make
make clean && TESTNUM=_close_count make 
```

I also reverted several changes and verified that the tests fail thereafter.

fmoralesc commented:
Thanks, @justinmk

glts commented:
@fmoralesc :+1:

Not all of the patch number changes in `version.c` made it in, can you fix this? Thanks.

fmoralesc commented:
@glts Opened a PR for it. I wonder what happened. (This thing just won't let me go! LOL)
